File: Data Structures & Algorithms/reorder-linked-list/submission-1.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def reorderList(self, head: Optional[ListNode]) -> None:
        # Step 1: Find midpoint
        slow, fast = head, head
        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next

        # Step 2: Reverse second half
        l2 = slow.next
        slow.next = None
        prev = None
        while l2:
            tmp = l2.next
            l2.next = prev
            prev = l2
            l2 = tmp
        l2 = prev  # l2 now points to reversed second half

        # Step 3: Merge two halves
        l1 = head
        while l2:
            tmp1, tmp2 = l1.next, l2.next
            l1.next = l2
            l2.next = tmp1
            l1, l2 = tmp1, tmp2
            

File: Data Structures & Algorithms/remove-node-from-end-of-linked-list/submission-1.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        int length = 0;
        ListNode* prev = nullptr;
        ListNode* curr = head;

        while (curr) {
            length++;
            curr = curr->next;
        }

        curr = head;

        for (int i = 0; i < length-n; i++) {
            prev = curr;
            curr = curr->next;
        }

        if (prev && curr) {
            prev->next = curr->next;
        } else if (curr) {
            head = curr->next;
        }

        curr->next = nullptr;
        curr = nullptr;

        return head;
    }
};
